File: problems/dynamic_programming/longest_common_subsequence.py
"""
LeetCode 1143. Longest Common Subsequence

Given two strings, return the length of their longest common subsequence
(not necessarily contiguous).

Approach: classic 2D DP. dp[i][j] = LCS length of text1[:i] and text2[:j].
If the last characters match, extend the diagonal (dp[i-1][j-1] + 1);
otherwise take the best of dropping a character from either string.

Time:  O(m * n)
Space: O(m * n) (can be reduced to O(min(m, n)) with row rolling)
"""


def longest_common_subsequence(text1: str, text2: str) -> int:
    m, n = len(text1), len(text2)
    dp = [[0] * (n + 1) for _ in range(m + 1)]
    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if text1[i - 1] == text2[j - 1]:
                dp[i][j] = dp[i - 1][j - 1] + 1
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
    return dp[m][n]


if __name__ == "__main__":
    assert longest_common_subsequence("abcde", "ace") == 3
    assert longest_common_subsequence("abc", "abc") == 3
    assert longest_common_subsequence("abc", "def") == 0
    print("All tests passed.")

File: solutions/problems/dynamicProgramming/longestCommonSubsequence.ts
/**
 * Approach: 2D DP. dp[i][j] = LCS length of text1[:i] and text2[:j]. If
 * the last characters match, extend the diagonal (dp[i-1][j-1] + 1);
 * otherwise take the best of dropping a character from either string.
 * Time: O(m * n). Space: O(m * n).
 */

export function longestCommonSubsequence(text1: string, text2: string): number {
  const m = text1.length;
  const n = text2.length;
  const dp: number[][] = Array.from({ length: m + 1 }, () => new Array(n + 1).fill(0));

  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      if (text1[i - 1] === text2[j - 1]) {
        dp[i][j] = dp[i - 1][j - 1] + 1;
      } else {
        dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
      }
    }
  }
  return dp[m][n];
}
